File: packages/wear/src/serialize.ts
// Canonical JSON for a WearDescriptor (§11.6 test 1: byte identical TS and Swift).
//
// Rules (the Swift port must emit exactly this):
//  - No whitespace anywhere.
//  - Keys in this fixed order, extra keys ignored:
//      {"version","seed","level","scratches","scuffZones","labelFade","edgeWear","dustAmount"}
//      scratch: {"surface","x","y","angle","length","depth"}
//      scuff:   {"surface","x","y","radius","intensity"}
//  - `version` is a plain base 10 integer.
//  - `seed` and `surface` are lower case ASCII identifiers, written in double
//    quotes with no escaping (validated, so none is ever needed).
//  - Every other number is on the 1e-6 grid and written from its integer
//    micro-unit value m as `${floor(m / 1e6)}.${m % 1e6 zero padded to 6}`,
//    e.g. 0 -> 0.000000, 350000 -> 0.350000, 123456789 -> 123.456789.
//    Never use the platform's float formatter (JS prints 1e-7, Swift 1e-07).

import { SURFACES } from './config';
import type { WearDescriptor } from './computeWear';

const MICRO = 1_000_000;
const LOWER_SEED_RE = /^[0-9a-f]{32}$/;

/** Formats a non negative integer micro-unit value with exactly six decimals. */
export function formatMicro(m: number): string {
  if (!Number.isSafeInteger(m) || m < 0) {
    throw new RangeError(`formatMicro expects a non negative integer, got ${m}`);
  }
  return `${Math.floor(m / MICRO)}.${String(m % MICRO).padStart(6, '0')}`;
}

/** Converts a descriptor number back to micro-units, refusing anything off the 1e-6 grid. */
function num(v: number, field: string): string {
  const m = Math.round(v * MICRO);
  if (typeof v !== 'number' || !Number.isFinite(v) || v < 0 || m / MICRO !== v) {
    throw new RangeError(`${field} is not a non negative value on the 1e-6 grid: ${v}`);
  }
  return formatMicro(m);
}

function surface(s: string, field: string): string {
  if (!(SURFACES as readonly string[]).includes(s)) throw new RangeError(`${field} is not a known surface: ${s}`);
  return `"${s}"`;
}

/** The canonical byte string for a descriptor. */
export function serializeDescriptor(d: WearDescriptor): string {
  if (!Number.isSafeInteger(d.version) || d.version < 1) throw new RangeError(`bad version: ${d.version}`);
  if (typeof d.seed !== 'string' || !LOWER_SEED_RE.test(d.seed)) throw new RangeError('seed must be 32 lower case hex characters');
  const scratches = d.scratches
    .map(
      (s, i) =>
        `{"surface":${surface(s.surface, `scratches[${i}].surface`)},"x":${num(s.x, 'x')},"y":${num(s.y, 'y')},` +
        `"angle":${num(s.angle, 'angle')},"length":${num(s.length, 'length')},"depth":${num(s.depth, 'depth')}}`,
    )
    .join(',');
  const scuffs = d.scuffZones
    .map(
      (s, i) =>
        `{"surface":${surface(s.surface, `scuffZones[${i}].surface`)},"x":${num(s.x, 'x')},"y":${num(s.y, 'y')},` +
        `"radius":${num(s.radius, 'radius')},"intensity":${num(s.intensity, 'intensity')}}`,
    )
    .join(',');
  return (
    `{"version":${d.version},"seed":"${d.seed}","level":${num(d.level, 'level')},` +
    `"scratches":[${scratches}],"scuffZones":[${scuffs}],` +
    `"labelFade":${num(d.labelFade, 'labelFade')},"edgeWear":${num(d.edgeWear, 'edgeWear')},` +
    `"dustAmount":${num(d.dustAmount, 'dustAmount')}}`
  );
}

File: packages/wear-swift/Sources/MyindWear/Serialize.swift
// Canonical serialization (README section 7). No whitespace, fixed key order, every non integer
// number written from its micro value as `floor(m / 1e6)` `.` `m % 1e6` zero padded to 6 digits.
// Never uses the platform float formatter.

/// Recover the micro value from a descriptor double (`v = m / 1e6`): `m = round(v * 1e6)`, and
/// throw unless `m >= 0`, `m` is a safe integer (< 2^53) and `m / 1e6 == v` (README section 7).
func toMicro(_ value: Double, field: String) throws -> Int64 {
  let scaled = value * 1_000_000
  guard scaled.isFinite else { throw WearError.nonCanonicalValue(field) }
  let rounded = scaled.rounded(.toNearestOrEven)   // JS Math.round differs only at .5, which is never exact here
  guard rounded >= 0, rounded < 9_007_199_254_740_992 else { throw WearError.nonCanonicalValue(field) }
  let m = Int64(rounded)
  guard Double(m) / 1_000_000 == value else { throw WearError.nonCanonicalValue(field) }
  return m
}

/// `123456789 -> 123.456789`, `0 -> 0.000000`, `350000 -> 0.350000`.
func formatMicro(_ m: Int64) -> String {
  precondition(m >= 0, "micro values are never negative")
  let whole = m / 1_000_000
  let frac = m % 1_000_000
  var fracText = String(frac)
  if fracText.count < 6 {
    fracText = String(repeating: "0", count: 6 - fracText.count) + fracText
  }
  return "\(whole).\(fracText)"
}

@inline(__always)
private func num(_ value: Double, _ field: String) throws -> String {
  formatMicro(try toMicro(value, field: field))
}

@inline(__always)
private func str(_ s: String) -> String {
  "\"" + s + "\""   // seed and surface are lower case ASCII; never need escaping
}

/// Throws `WearError.nonCanonicalValue` if a value is not an exact micro quotient.
public func serializeDescriptor(_ d: WearDescriptor) throws -> String {
  var out = "{"
  out += "\"version\":\(d.version)"
  out += ",\"seed\":" + str(d.seed)
  out += ",\"level\":" + (try num(d.level, "level"))
  out += ",\"scratches\":["
  for (i, s) in d.scratches.enumerated() {
    if i > 0 { out += "," }
    out += "{\"surface\":" + str(s.surface.rawValue)
    out += ",\"x\":" + (try num(s.x, "scratches[\(i)].x"))
    out += ",\"y\":" + (try num(s.y, "scratches[\(i)].y"))
    out += ",\"angle\":" + (try num(s.angle, "scratches[\(i)].angle"))
    out += ",\"length\":" + (try num(s.length, "scratches[\(i)].length"))
    out += ",\"depth\":" + (try num(s.depth, "scratches[\(i)].depth"))
    out += "}"
  }
  out += "],\"scuffZones\":["
  for (i, z) in d.scuffZones.enumerated() {
    if i > 0 { out += "," }
    out += "{\"surface\":" + str(z.surface.rawValue)
    out += ",\"x\":" + (try num(z.x, "scuffZones[\(i)].x"))
    out += ",\"y\":" + (try num(z.y, "scuffZones[\(i)].y"))
    out += ",\"radius\":" + (try num(z.radius, "scuffZones[\(i)].radius"))
    out += ",\"intensity\":" + (try num(z.intensity, "scuffZones[\(i)].intensity"))
    out += "}"
  }
  out += "],\"labelFade\":" + (try num(d.labelFade, "labelFade"))
  out += ",\"edgeWear\":" + (try num(d.edgeWear, "edgeWear"))
  out += ",\"dustAmount\":" + (try num(d.dustAmount, "dustAmount"))
  out += "}"
  return out
}

/// The canonical bytes (UTF-8 of `serializeDescriptor`; the output is pure ASCII).
public func serializeDescriptorBytes(_ d: WearDescriptor) throws -> [UInt8] {
  Array(try serializeDescriptor(d).utf8)
}
